llvm/lib/Target/AMDGPU/VINTERPInstructions.td

```diff
@@ -181,9 +181,43 @@ multiclass VInterpF16Pat <SDPatternOperator op, Instruction inst,
   def : VInterpF16Pat<op, inst, dst_type, 1, high_pat>;
 }
 
+class VInterpF16Pat_t16 <SDPatternOperator op, Instruction inst,
+                     ValueType dstVT, bit high, bit isP2> : GCNPat <
+   (dstVT (op
+      (VINTERPMods f32:$src0, i32:$src0_modifiers),
+      (VINTERPMods f32:$src1, i32:$src1_modifiers),
+      (VINTERPMods f32:$src2, i32:$src2_modifiers),
+      !if(high, (i1 -1), (i1 0)))),
+    (inst $src0_modifiers,
+          (f16 (EXTRACT_SUBREG VGPR_32:$src0, !if(high, hi16, lo16))),
```

Comment:
How does arbitrarily extracting from a float work?

Reply:
Yes

Reply:
I see. I think this is not extracting from a 32bits float to a 16bits float. The input is always a 16bits float and this is just selecting the high half or low half of a 32 bit register

Reply:
So the marked types are wrong, they shouldn't be f32

Reply:
Hi Matt. I think the input types are depending on the intrinsic. From this test https://github.com/llvm/llvm-project/pull/116702/files#diff-67c39b2811e166871c5e79f74f33221b90c06a20645efe1745e294f99cb3bec8R123 it seems we are passing a 32 bits float here as the input

Reply:
After looking at the documentation for these intrinsics, I do not understand why they are using float. src0 and src2 should be using `<2 x half>`?

Reply:
It seems there is a historical reason. 

https://reviews.llvm.org/D127756 The same question is discussed here and the reason might be this:
```
Wouldn't it be more natural to declare p and p0 to be llvm_v2f16_ty?

I think so, but it would require coordinated changes in Mesa and LLPC, and we should probably also provide a v2f16 version of llvm.amdgcn.lds.param.load which currently always returns a float.
```

Reply:
The historical reason is it was just wrong to begin with

Reply:
Hi Matt. Regarding the actual implementation I think @Sisyph knows better than I do. He is in vacation right now and he will come back next week. We can dicuss the next step with some input from him. Thanks!

Reply:
Yes, I agree with @arsenm that these should have been 2 x half to begin with. However, I would be in favor of landing this patch as is, and adding a comment and filing an LLVM issue to change the intrinsic type. Since this is working for the current frontends it does not seem urgent to change the instrinsic.